File: server/app/modules/game/interfaces/tile.ts
export interface Tile {
    type: TileType;
    orientation?: string;
    state?: string;
}

export enum TileType {
    Snow = 'snow',
    Tree = 'tree',
    Stone = 'stone',
    Ice = 'ice',
    Water = 'water',
    Door = 'door',
    Wall = 'wall',
    Corner = 'corner',
    Intersection = 'intersection',
    Trap = 'trap',
    TeleportPad = 'teleportPad',
}

export enum MoveCosts {
    Ice = 0,
    Snow = 1,
    Water = 2,
    WaterWithBoots = 1,
    Tree = Infinity,
    Stone = Infinity,
    Door = 1,
    Wall = Infinity,
    Corner = Infinity,
    Intersection = Infinity,
    Trap = 1,
    TeleportPad = 1,
}

File: client_leger/lib/features/game_session/domain/models/tile.dart
import '../../core/enums/tile_type.dart';

sealed class Tile {
  const Tile();

  TileType get type;

  factory Tile.fromType(TileType type, [TileState? state]) => switch (type) {
    TileType.snow => const SnowTile(),
    TileType.tree => const TreeTile(),
    TileType.stone => const StoneTile(),
    TileType.ice => const IceTile(),
    TileType.water => const WaterTile(),
    TileType.door => DoorTile(state ?? TileState.closed),
    TileType.wall => const WallTile(),
    TileType.corner => const CornerTile(),
    TileType.intersection => const IntersectionTile(),
  };
}

final class SnowTile extends Tile {
  const SnowTile();
  @override
  TileType get type => TileType.snow;
}

final class TreeTile extends Tile {
  const TreeTile();
  @override
  TileType get type => TileType.tree;
}

final class StoneTile extends Tile {
  const StoneTile();
  @override
  TileType get type => TileType.stone;
}

final class IceTile extends Tile {
  const IceTile();
  @override
  TileType get type => TileType.ice;
}

final class WaterTile extends Tile {
  const WaterTile();
  @override
  TileType get type => TileType.water;
}

final class DoorTile extends Tile {
  final TileState state;
  const DoorTile(this.state);
  @override
  TileType get type => TileType.door;

  DoorTile toggle() =>
      DoorTile(state == TileState.opened ? TileState.closed : TileState.opened);
}

final class WallTile extends Tile {
  const WallTile();
  @override
  TileType get type => TileType.wall;
}

final class CornerTile extends Tile {
  const CornerTile();
  @override
  TileType get type => TileType.corner;
}

final class IntersectionTile extends Tile {
  const IntersectionTile();
  @override
  TileType get type => TileType.intersection;
}
